geopandas/geodataframe.py

```diff
@@ -1515,6 +1543,10 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
         result = super().apply(
             func, axis=axis, raw=raw, result_type=result_type, args=args, **kwargs
         )
+        # Check if last geometry column name is lost and re-attach
+        # (should only be for pandas 1.3 and earlier)
+        if isinstance(result, GeoDataFrame) and result._geometry_column_name is None:
+            result._geometry_column_name = self._geometry_column_name
```

Comment:
The only reason the associated tests in test_op_output_types passed for old versions of pandas is that the geometry column assigned by default "geometry", happened to also be the one missing. (I presume a missing finalize call but haven't dug into).

Note the below condition is intentionally an if and not an elif as it applies to both DataFrames and GeoDataFrames

Reply:
Could we add a `not compat.PANDAS_GE_14 and ..`? Then we are sure to already run and test it without the workaround for recent versions of pandas, and we will also more easily remember to remove this (if we remove that variable)